fedimint-core/src/db/notifications.rs

```diff
@@ -0,0 +1,236 @@
+        }
+    }
+}
+
+impl NotifyQueue {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn add<K>(&mut self, key: &K)
+    where
+        K: Hash,
+    {
+        let mut hasher = DefaultHasher::new();
+        key.hash(&mut hasher);
+        let hash_value = hasher.finish();
+        let slot_idx = (hash_value % (self.buckets.len() as u64)) as usize;
+        self.buckets.set(slot_idx, true);
+    }
+}
+
+/// Wrapper to implements `add_notification_key`
+pub struct NotifyingTransactionWrapper<'a> {
+    // TODO: try removing Box
+    dbtx: Box<dyn ISingleUseDatabaseTransaction<'a>>,
+    // notifications to be submited after commit
+    notify_queue: Option<NotifyQueue>,
+    notifications: &'a Notifications,
+}
+
+impl<'a> NotifyingTransactionWrapper<'a> {
+    pub fn new(
+        dbtx: Box<dyn ISingleUseDatabaseTransaction<'a>>,
+        notifications: &'a Notifications,
+    ) -> Self {
+        Self {
+            dbtx,
+            notify_queue: Some(NotifyQueue::new()),
+            notifications,
+        }
+    }
+}
+
+#[apply(async_trait_maybe_send!)]
+impl<'a> ISingleUseDatabaseTransaction<'a> for NotifyingTransactionWrapper<'a> {
+    async fn raw_insert_bytes(&mut self, key: &[u8], value: Vec<u8>) -> Result<Option<Vec<u8>>> {
+        self.dbtx.raw_insert_bytes(key, value).await
+    }
+
+    async fn raw_get_bytes(&mut self, key: &[u8]) -> Result<Option<Vec<u8>>> {
+        self.dbtx.raw_get_bytes(key).await
+    }
+
+    async fn raw_remove_entry(&mut self, key: &[u8]) -> Result<Option<Vec<u8>>> {
+        self.dbtx.raw_remove_entry(key).await
+    }
+
+    async fn raw_find_by_prefix(&mut self, key_prefix: &[u8]) -> Result<PrefixStream<'_>> {
+        self.dbtx.raw_find_by_prefix(key_prefix).await
+    }
+
+    async fn raw_remove_by_prefix(&mut self, key_prefix: &[u8]) -> Result<()> {
+        self.dbtx.raw_remove_by_prefix(key_prefix).await
+    }
+
+    async fn commit_tx(&mut self) -> Result<()> {
+        self.dbtx.commit_tx().await?;
+        self.notifications.submit_queue(
+            self.notify_queue
+                .take()
+                .expect("commit must be called only once"),
+        );
+        Ok(())
+    }
+
+    async fn rollback_tx_to_savepoint(&mut self) -> Result<()> {
+        self.dbtx.rollback_tx_to_savepoint().await
+    }
+
+    async fn set_tx_savepoint(&mut self) -> Result<()> {
+        self.dbtx.set_tx_savepoint().await
+    }
+
+    fn add_notification_key(&mut self, key: &[u8]) -> Result<()> {
+        self.notify_queue
+            .as_mut()
+            .context("can not call add_notification_key after commit")?
+            .add(&key);
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::time::Duration;
+
+    use futures::Future;
+
+    use super::*;
+
+    async fn future_returns_shortly<F: Future>(sub: F) -> bool {
+        tokio::time::timeout(Duration::from_millis(10), sub)
+            .await
+            .is_ok()
+    }
```

Comment:
Nit: This exact same function is defined twice. Perhaps we could move it to some kind of shared test utils crate.